docs/community/tutorials-contest/2022/red-dev-ecrecover-x-chain/README.md

```diff
@@ -0,0 +1,494 @@
+
+func allZero(b []byte) bool {
+	for _, byte := range b {
+		if byte != 0 {
+			return false
+		}
+	}
+	return true
+}
+
+// getXChainECRecover returns an execution function that reads the input and return the input from the given [precompileAddr].
+// The execution function parses the input into a string and returns the string
+func getXChainECRecover(precompileAddr common.Address) RunStatefulPrecompileFunc {
+	return func(evm PrecompileAccessibleState, callerAddr common.Address, addr common.Address, input []byte, suppliedGas uint64, readOnly bool) (ret []byte, remainingGas uint64, err error) {
+		if remainingGas, err = deductGas(suppliedGas, XChainECRecoverCost); err != nil {
+			return nil, 0, err
+		}
+
+		const ecRecoverInputLength = 128
+
+		input = common.RightPadBytes(input, ecRecoverInputLength)
+
+		// "input" is (hash, v, r, s), each 32 bytes
+		// but for ecrecover we want (r, s, v)
+
+		r := new(big.Int).SetBytes(input[64:96])
+		s := new(big.Int).SetBytes(input[96:128])
+		v := input[63]
+
+		// tighter sig s values input homestead only apply to tx sigs
+		if !allZero(input[32:63]) || !crypto.ValidateSignatureValues(v, r, s, false) {
+			return nil, remainingGas, nil
+		}
+
+		// We must make sure not to modify the 'input', so placing the 'v' along with
+		// the signature needs to be done on a new allocation
+
+		sig := make([]byte, 65)
+		copy(sig, input[64:128])
+		sig[64] = v
+
+		// v needs to be at the end for libsecp256k1
+		pubk, err := crypto.SigToPub(input[:32], sig)
+		publicKey := crypto.CompressPubkey(pubk)
+
+		// make sure the public key is a valid one
+		if err != nil {
+			return nil, remainingGas, nil
+		}
+
+		sha := sha256.Sum256(publicKey)
+		ripemd := ripemd160.New()
+		ripemd.Write(sha[:])
+		ripe := ripemd.Sum(nil)
+
+		conv, err := bech32.ConvertBits(ripe, 8, 5, true)
+		if err != nil {
+			log.Info("Error:", err)
+		}
+		encoded, err := bech32.Encode("fuji", conv)
+		xchain := "X-" + encoded
+
+		if err != nil {
+			log.Info("Error:", err)
+		}
+
+		return []byte(xchain), remainingGas, nil
+	}
+}
+
+// createXChainECRecoverPrecompile returns a StatefulPrecompiledContract with R/W control of an allow list at [precompileAddr] and a native coin minter.
+func createXChainECRecoverPrecompile(precompileAddr common.Address) StatefulPrecompiledContract {
+	funcGetXChainECRecover := newStatefulPrecompileFunction(xChainECRecoverReadSignature, getXChainECRecover(precompileAddr))
+
+	// Construct the contract with no fallback function.
+	contract := newStatefulPrecompileWithFunctionSelectors(nil, []*statefulPrecompileFunction{xChainECRecover,funcGetXChainECRecover})
+	return contract
+}
+```
+
+(Please note that we are still working to repair one final defect in this example contract. After computing the correct X-Chain address of the signer, it does not yet return this value to the calling dApp due to an issue concerning data types. We expect to have this resolved soon.)
```

Comment:
Has this issue been resolved?

Reply:
Not yet @Julian-dev28 . I've been asking Aaron Buchwald for some ideas today, but we're not there yet. Aratrika will be trying some things in the next 12 hours.

Reply:
Great!

Please ping me for review when you have implemented this so I can test locally.

Thank you

Reply:
@Julian-dev28 It's going to be 24 more hours. We have a simple precompile that works, and then we have our ecrecover-x-chain precompile that throws the error, and the solution is somewhere between the two, the simple one that works and the complex one that does not. I think we'll have it figured out soon. However, Aratrika is in India, so it's time for sleep, and also Tamil had to take some personal time this week. Meanwhile I'll remove the one large code snippit from the tutorial in order to meet your guidelines.

Reply:
@kitpub 

Understandable. Thank you for keeping me in the loop.
Looking forward to trying out the new solution when its shipped.